File: mobile/lib/domain/usecases/select_course_use_case.dart
import '../repositories/i_course_repository.dart';
import '../repositories/i_learner_repository.dart';
import '../entities/course.dart';
import '../../core/errors/domain_error.dart';

class SelectCourseUseCase {
  final ICourseRepository courseRepository;
  final ILearnerRepository learnerRepository;

  const SelectCourseUseCase(this.courseRepository, this.learnerRepository);

  Future<Course> execute(String courseId) async {
    final course = await courseRepository.getCourseById(courseId);
    if (course == null) {
      throw EntityNotFoundError('Kurs', courseId);
    }

    final profile = await learnerRepository.getProfile();
    await learnerRepository.updateProfile(profile.copyWith(selectedCourseId: courseId));
    return course;
  }
}

File: src/domain/usecases/SelectCourseUseCase.ts
import { ICourseRepository } from '../repositories/ICourseRepository';
import { ILearnerRepository } from '../repositories/ILearnerRepository';
import { Course } from '../entities/Course';
import { EntityNotFoundError } from '@/core/errors/DomainError';

export class SelectCourseUseCase {
  constructor(
    private courseRepo: ICourseRepository,
    private learnerRepo: ILearnerRepository
  ) {}

  async execute(courseId: string): Promise<Course> {
    const course = await this.courseRepo.getCourseById(courseId);
    if (!course) {
      throw new EntityNotFoundError('Kurs', courseId);
    }

    await this.learnerRepo.updateProfile({ selectedCourseId: courseId });
    return course;
  }
}
